src/ucs/datastruct/bitmap.h

```diff
@@ -126,71 +289,120 @@ _UCS_BITMAP_DECLARE_TYPE(256)
  * @param _bitmap Clear all bits in this bitmap
  */
 #define UCS_BITMAP_CLEAR(_bitmap) \
-    memset((_bitmap)->bits, 0, sizeof((_bitmap)->bits))
+    memset((_bitmap).bits, 0, sizeof((_bitmap).bits))
 
 
 /**
- * Iterate over all set (1) bits of a given bitmap
+ * Find the index of the first bit set to 1 in a given bitmap
  *
- * @param _bitmap Iterate over bits of this bitmap
- * @param _index  Bit index (global offset - relative to the whole bitmap)
+ * @param _bitmap Look for the first bit in this bitmap
  */
-#define UCS_BITMAP_FOR_EACH_BIT(_bitmap, _index) \
-    for (_index = ucs_ffs64_safe((_bitmap)->bits[0]); \
-         _index < ucs_static_array_size((_bitmap)->bits) * UCS_BITMAP_BITS_IN_WORD; \
-         _index = _UCS_BITMAP_WORD_INDEX0(_index) + \
-                  ucs_ffs64_safe(_UCS_BITMAP_WORD((_bitmap), (_index)) & \
-                                 _UCS_BITMAP_GET_NEXT_BIT(_index))) \
-        if (UCS_BITMAP_GET((_bitmap), (_index)))
+#define UCS_BITMAP_FFS(_bitmap) \
+    ({ \
+        size_t _word_index, _index; \
+        _UCS_BITMAP_FOR_EACH_WORD(_bitmap, _word_index) { \
+            _index = ucs_ffs64_safe(_UCS_BITMAP_WORD(_bitmap, _word_index)) + \
+                     _word_index * UCS_BITMAP_BITS_IN_WORD; \
+            if (_index % UCS_BITMAP_BITS_IN_WORD) { \
+                break; \
+            } \
+        } \
+        _index; \
+    })
 
 
 /**
- * Perform bitwise NOT of a bitmap
+ * Return the number of bits set to 1 in a given bitmap
  *
- * @param _bitmap Negate this bitmap
+ * @param _bitmap Check bits number in this bitmap
+ *
+ * @return Number of bits set to 1
  */
-#define UCS_BITMAP_NOT(_bitmap) \
+#define UCS_BITMAP_POPCOUNT(_bitmap) \
     ({ \
-        int _word; \
-        for (_word = 0; _word < ucs_static_array_size((_bitmap)->bits); _word++) \
-            (_bitmap)->bits[_word] = ~(_bitmap)->bits[_word]; \
+        size_t _word_index = 0, _popcount = 0; \
+        _UCS_BITMAP_FOR_EACH_WORD(_bitmap, _word_index) { \
+            _popcount += ucs_popcount(_UCS_BITMAP_WORD(_bitmap, _word_index)); \
+        } \
+        _popcount; \
     })
 
 
-#define _UCS_BITMAP_OP(_bitmap1, _bitmap2, _op) \
+/**
+ *  Returns the number of bits set to 1 in a given bitmap,
+ *  up to a particular index
+ *
+ * @param _bitmap Check bits number in this bitmap
+ * @param _index  Check bits up to this index
+ *
+ * @return Number of bits set to 1
+ */
+#define UCS_BITMAP_POPCOUNT_UPTO_INDEX(_bitmap, _index) \
```

Comment:
maybe implement UCS_BITMAP_POPCOUNT by using UCS_BITMAP_POPCOUNT_UPTO_INDEX?

Reply:
I didn't do it intentionally - `UCS_BITMAP_POPCOUNT` is simpler and faster, because the assumptions it makes.

Reply:
maybe compiler could eliminate extra check if "upto index" was constant